app/upgrades/v17/erc20_test.go

```diff
@@ -0,0 +1,191 @@
+	}{
+		{
+			name:    "pass - no token pairs in ERC20 keeper",
+			expPass: true,
+			postCheck: func() {
+				requireActiveDynamicPrecompiles(t, network.GetContext(), network.App.EvmKeeper, []string{})
+			},
+		},
+		{
+			name: "pass - native token pair",
+			malleate: func() {
+				network.App.Erc20Keeper.SetTokenPair(network.GetContext(), nativeTokenPair)
+			},
+			expPass: true,
+			postCheck: func() {
+				// Check that the precompile was registered
+				activeDynamicPrecompiles := network.App.EvmKeeper.GetParams(network.GetContext()).ActiveDynamicPrecompiles
+				require.Contains(t, activeDynamicPrecompiles, nativeTokenPair.GetERC20Contract().String(), "expected precompile to be registered")
+
+				// Check that the precompile is set as active
+				requireActiveDynamicPrecompiles(
+					t, network.GetContext(), network.App.EvmKeeper,
+					[]string{nativeTokenPair.Erc20Address},
+				)
+			},
+		},
+		{
+			name: "pass - IBC token pair",
+			malleate: func() {
+				network.App.Erc20Keeper.SetTokenPair(network.GetContext(), ibcTokenPair)
+			},
+			expPass: true,
+			postCheck: func() {
+				// Check that the precompile was registered
+				activeDynamicPrecompiles := network.App.EvmKeeper.GetParams(network.GetContext()).ActiveDynamicPrecompiles
+				require.Contains(t, activeDynamicPrecompiles, ibcTokenPair.GetERC20Contract().String(), "expected precompile to be registered")
+
+				// Check that the precompile is set as active
+				requireActiveDynamicPrecompiles(
+					t, network.GetContext(), network.App.EvmKeeper,
+					[]string{ibcTokenPair.Erc20Address},
+				)
+			},
+		},
+		{
+			name: "pass - external token pair is skipped",
+			malleate: func() {
+				network.App.Erc20Keeper.SetTokenPair(network.GetContext(), externalTokenPair)
+				network.App.Erc20Keeper.SetTokenPair(network.GetContext(), otherNativeTokenPair)
+			},
+			expPass: true,
+			postCheck: func() {
+				// Check that active precompiles are unchanged
+				requireActiveDynamicPrecompiles(
+					t, network.GetContext(), network.App.EvmKeeper,
+					[]string{otherNativeTokenPair.Erc20Address},
+				)
+			},
+		},
+		{
+			// TODO: this is not failing anymore in the new implementation? Is that what we want?
+			name: "no-op - already registered precompile",
+			malleate: func() {
+				network.App.Erc20Keeper.SetTokenPair(network.GetContext(), nativeTokenPair)
+				network.App.Erc20Keeper.SetTokenPair(network.GetContext(), otherNativeTokenPair)
+
+				tokenPrecompile, err := erc20precompile.NewPrecompile(
+					nativeTokenPair, network.App.BankKeeper, network.App.AuthzKeeper, network.App.TransferKeeper,
+				)
+				require.NoError(t, err, "expected no error creating precompile")
+
+				// NOTE: We are adding this in the malleate function, so it is already present before
+				// the real test case.
+				err = network.App.EvmKeeper.AddDynamicPrecompiles(network.GetContext(), tokenPrecompile)
+				require.NoError(t, err, "expected no error adding precompile")
+			},
```

Comment:
The case where a precompile is already registered is not causing an error anymore in this new implementation - I'm unsure that we actually want this? 👀

Reply:
i think we should return an error.